<a href="https://colab.research.google.com/github/111Vidya/Python_fundamentals/blob/main/Mar_26_Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Univariate Feature Selection or Chisquare method for feature selection

In [1]:
# Feature Extraction with Univariate Statistical Tests (Chi-squared for classification)
from pandas import read_csv
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
# load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# feature extraction
test = SelectKBest(score_func=chi2, k=4) #4 best features using chi2 method from selectKBest
fit = test.fit(X, Y)
# summarize scores
set_printoptions(precision=3)
print(fit.scores_)
features = fit.transform(X)


#For regression: f_regression, mutual_info_regression
#For classification: chi2, f_classif, mutual_info_classif

[ 111.52  1411.887   17.605   53.108 2175.565  127.669    5.393  181.304]


2175.565 - test, 1411.88-Plas, 181.30-age,127.66-mass 4 import IV features.
Class is TV

#### Recursive Feature Elimination

In [2]:
# Feature Extraction with RFE
from pandas import read_csv
from sklearn.feature_selection import RFE # library for rFE
from sklearn.linear_model import LogisticRegression
# load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# feature extraction
model = LogisticRegression(max_iter=400)



In [8]:
#select top 3 features
rfe = RFE(model, n_features_to_select=3, step=1)
fit = rfe.fit(X, Y)

In [9]:
#Num Features:
fit.n_features_

3

In [10]:
#Selected Features:
fit.support_

array([False,  True, False, False, False,  True,  True, False])

In [11]:
# Feature Ranking:
fit.ranking_

array([4, 1, 3, 6, 5, 1, 1, 2])

Rank 1 in thid case is importat features. all the ones should be chosen

#### Feature Importance using Decision Tree

In [7]:
# Feature Importance with Extra Trees Classifier
from pandas import read_csv
from sklearn.tree import  DecisionTreeClassifier # library for DT
# load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# feature extraction
model = DecisionTreeClassifier()
model.fit(X, Y)
print(model.feature_importances_)

[0.06  0.316 0.084 0.012 0.037 0.25  0.127 0.114]


High values are the important features . 0.316-plas,0.25-mass,0.127-pedi,0.114-age